### PR TITLE
[Hub Generated] Review request for Microsoft.DigitalTwins to add version stable/2022-10-31

### PR DESCRIPTION
**Pull request by dejimarquis**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/dejimarquis/azure-rest-api-specs/release-digitaltwins-Microsoft.DigitalTwins-2022-10-31...main/).
## ARM API Information (Control Plane)

Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.  ---- **10/31/2022**
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month. --- **10/24/2022**
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[x] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (zizw123) [inline specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:1317]: >SystemAssigned,UserAssigned [](http://example.com/codeflow?start=13&length=27)

This means both user and system assigned? Curious if there's any thought on providing one single for this case

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (zizw123) [inline specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:1323]: >SystemAssigned [](http://example.com/codeflow?start=94&length=14)

This need to be updated

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (efriesner) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:1317]: This is following the contract set by the Managed Identity team and what other RPs have defined in their swagger for User Assigned Identities: https://dev.azure.com/msazure/One/_releaseProgress?_a=release-pipeline-progress&releaseId=7864354

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (dejimarquis) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json]: Resolved this

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (raosuhas) [inline specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:1314]: Any reason you are not using the common types : 

[https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v3/managedidentity.json](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v3/managedidentity.json)

Comment (raosuhas) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:1314]: All other issues could be avoided if you were using common types from the beginning

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (raosuhas) [inline specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:2119]: >identity [](http://example.com/codeflow?start=9&length=8)

Identity is a top level property not meant to be under properties Are you doing some custom implementation ?

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (sjiherzig) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:1314]: If we start using these now, wouldn't that cause a breaking change? and not only here, but also in management SDKs?)

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (efriesner) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:2119]: Yes, this is a custom implementation. This isn't the same as the 'identity' property for DigitalTwinsInstances - TimeSeriesDatabaseConnection is a proxy resource

Comment (raosuhas) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:2119]: Why do you need the custom implementation ? Proxy resources support MSI as explained here : 
https://armwiki.azurewebsites.net/authorization/managed_identities/MSIOnboardingProxyResource.html

Full details here : 
https://armwiki.azurewebsites.net/authorization/managed_identities/MSIOnboardingSteps.html

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (raosuhas) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:1314]: Identity is a standard contract across all azure resources. Unless you are using any additional custom field it should not cause any breaking change

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (efriesner) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:2119]: The intention is not to assign a new identity to a proxy resource, but rather choose an identity associated with a DigitalTwinsInstance, for aunthentication to a customer owned resource. This is similar to what IoTHub has done [here](https://github.com/Azure/azure-rest-api-specs/blob/450b1618259d8b170b5cfcb37866d3fc5747b564/specification/iothub/resource-manager/Microsoft.Devices/stable/2021-07-02/iothub.json#L2518). We already support this functionality for system assigned identities, this change is now extending the identity based authentication to support user assigned identities.

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (efriesner) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json]: We tried updating to the common type but unfortunately it it will cause breaking changes in our SDKs. We would prefer to stick with the existing types

**Comment by dejimarquis** [comment]
@raosuhas @changlong-liu the breaking change error is a false positive. Can you please correct?

**Comment by dejimarquis** [comment]
@raosuhas @changlong-liu the breaking change error is a false positive. Can you please, correct?

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (raosuhas) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:2119]: Got it , thanks for explaining. Can you please rename the "ManagedIdentity" to something like "ManagedIdentityReference" so that this is clear for future reviewers ? Signing off on this otherwise , please fix before merging.

**Comment by changlong-liu** [comment]
Hi @qiaozha , would you help to take a look whether reported JS breaking change blocks this PR? Thanks.

**Comment by changlong-liu** [comment]
Hi @dejimarquis , since the "API Readiness Check" fail, would you help check whether the service has been deployed? Generally, swagger need to get merged after service is ready.

File: specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json
Comment (dejimarquis) [reply specification/digitaltwins/resource-manager/Microsoft.DigitalTwins/stable/2022-10-31/digitaltwins.json:2119]: This is resolved!

**Comment by dejimarquis** [comment]
> 

@changlong-liu We haven't deployed yet. The service should be fully deployed by 10/31

**Comment by dejimarquis** [comment]
> Hi @dejimarquis , since the "API Readiness Check" fail, would you help check whether the service has been deployed? Generally, swagger need to get merged after service is ready.

@changlong-liu  We haven't deployed yet. The service should be fully deployed by 10/31

**Comment by changlong-liu** [comment]
> > Hi @dejimarquis , since the "API Readiness Check" fail, would you help check whether the service has been deployed? Generally, swagger need to get merged after service is ready.
> 
> @changlong-liu We haven't deployed yet. The service should be fully deployed by 10/31

Hi @dejimarquis , I added label for approval. We may come back to merge after service deployed.

**Review by efriesner — APPROVED** [review]

**Review by sjiherzig — APPROVED** [review]

**Pull request merged**
